### PR TITLE
binutils: bump to 2.38 + test more

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **binutils/2.38**

Closes https://github.com/conan-io/conan-center-index/issues/10385
Closes https://github.com/conan-io/conan-center-index/issues/10365

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/binutils/all/conanfile.py [hunk 1]
```diff
@@ -1,70 +1,154 @@
...
+
+    def config_options(self):
+        del self.settings.compiler.cppstd
+        del self.settings.compiler.libcxx
+        self.options.target_arch = str(getattr(self, "settings_target", self.settings).arch)
+        self.options.target_os = str(getattr(self, "settings_target", self.settings).os)
 
     def validate(self):
-        if self.settings.os not in ["Linux", "FreeBSD"]:
+        if self.settings.os not in ("Linux", "FreeBSD"):
```
Comment (madebr) [inline recipes/binutils/all/conanfile.py:44]: @hlewin

I've kept this validation.
This line means the cci recipe currently supports building binutils on Linux/FreeBSD.

But the recipe allows you to build binutils that targets Windows/FreeBSD/Macos/...

In the test_package, some architectures are tested.
I've verified that the executables targetting: x86@Linux, x86_64@Linux, x86@Windows and x86_64@Windows work (using wine)
I did not verify the armv8@Linux executable (yet)

**Comment by madebr** [comment]
@SSE4 
Would you mind having a quick glance why armv8@Macos is failing? x86_64@Macos seems to build just fine.
It has some problems with the archiver (`'aarch64-apple-darwin-ar`).

Binutils does not support the assembler (gas) and the linker (ld) on Macos.
in your opinion, is binutils useful on Macos?
https://github.com/Homebrew/homebrew-core/issues/29177

**Comment by SSE4** [comment]
> Binutils does not support the assembler (gas) and the linker (ld) on Macos. in your opinion, is binutils useful on Macos? If you're having doubts, we can just disable it.

probably, it's not useful. let's disable it until someone actuallly requests.

File: recipes/binutils/all/conanfile.py [hunk 2]
```diff
@@ -1,70 +1,160 @@
...
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _settings_build(self):
+        return getattr(self, "settings_build", self.settings)
+
+    @property
+    def _settings_target(self):
+        return getattr(self, "settings_target", None) or self.settings
```
Comment (jgsogo) [inline recipes/binutils/all/conanfile.py:43]: I don't think this is going to work... but I don't really know how `binutils` works. So, please, read my lines in the context of a cross-compiler (Android NDK, emsdk,...) and, if this is not the case for `binutils`, just ignore all the following.

(After reading previous paragraph, continue at your own risk 😄 )

Here (**while creating the package**) we are assigning to `option.target_os` the value of the _host_ os, which means that it is targeting the same OS where it runs. We built the package for the _host_ OS and the value of `option.target_os` is that same OS. This is essentially wrong is the purpose is to target a different OS from the one where `binutils` is running (~crosscompiling scenario).

When I suggest using an option it is because the binary (the files in the Conan package) is different depending on the target system, like containing a different sysroot... libraries pre-compiled for that target system. With this approach, at the time of generating the package the creator needs to provide a value explicitly:

 * **`pr-linux-to-windows`**: profile building Linux package that targets Windows
   ```ini
   [settings]
   os=Linux
   [options]
   binutils:target_os=Windows
   ```
   This would generate a `binutils` package that runs in Linux and contains all the DLLs required to target Windows:
   ```
   conan create binutils/xxxx --profile:host=pr-linux-to-windows
   ```

 * **`pr-linux-to-macos`**: profile building Linux package that targets Windows
   ```ini
   [settings]
   os=Linux
   [options]
   binutils:target_os=macos
   ```
   This would generate a `binutils` package that runs in Linux and contains all the DYLIBs required to target Macos:
   ```
   conan create binutils/xxxx --profile:host=pr-linux-to-macos
   ```
   
Both packages are different because they have different artifacts inside. As the settings are equal (both are to be run in Linux) we can only have different packageIDs using an option. We use that option (explicitly set at the time of building) to choose the artifacts to include in the resulting package.

---

Then we move to the consumption stage. In order for Conan to select the proper package, the consumer needs to provide settings and options that compute the desired packageID. As this is used as a build-requires in the build context, we can use the same profiles we used to generate those packages: 
 * `--profile:build=pr-linux-to-windows` will select the package that runs in Linux and is capable of targeting Windows
 * `--profile:build=pr-linux-to-macos` will select the package that runs in Linux and is capable of targeting Windows

And the consumer needs to provide the profile for the _host_ context as well:
 * if targeting windows, the _host_ profile should match Windows: `--profile:host=windows`
 * if targeting macos, the _host_ profile should match Macos: `--profile:host=macos`

**All of this is implicit knowledge on the consumer side**, the target architecture at consumption time used by `binutils` in the _build_ context needs to match the _host_ context. **Our recipe can only inform/raise** the user about a mismatch, but it cannot modify settings (it would be chicken-egg problem):

```python
def validate(self):
    if hasattr(self, 'settings_target'):  # We can only check when USED as build-requires
       if self.options.target_os != self.settings_target.os:
          raise ConanInvalid.... "Mismatch between target architecture and the one you are building your packages for"
``` 

Please, forgive me if `binutils` is not something like a cross-compiler.

Comment (madebr) [reply recipes/binutils/all/conanfile.py:43]: `binutils` is a cross-compiler, so build/host/target can be all different from each other

In this recipe, I basically do:
```python
class BinutilsConan(ConanFile):
    @property
    def _settings_target(self):
        return getattr(self, "settings_target", None) or self.settings
        
    def config_options(self):
        self.options.target_arch = str(self._settings_target.arch)
        self.options.target_os = str(self._settings_target.os)
        
    def validate(self):
        if self.options.target_arch not in self.settings.arch.values_range:
            raise ConanInvalidConfiguration(f"target_arch={self.options.target_arch} is invalid (possibilities={self.settings.arch.values_range})")
        if self.options.target_os not in self.settings.os.values_range:
            raise ConanInvalidConfiguration(f"target_os={self.options.target_os} is invalid (possibilities={self.settings.os.values_range})")
```

By assigning `config_option.target_{arch,os}` in `config_options` (overridable by the user), isn't conan assigning different package ids?

I created the following table:


| build    |   host    |   target    | package id                                 | command
| -------- | --------- | ----------- | ------------------------------------------ | ------------
| Linux    | Linux     | Linux       | `d0c627b96beba6c4f548b2e66247eece9701544a` | `conan create . binutils/2.38@`
| Linux    | Linux     | Linux       | `d0c627b96beba6c4f548b2e66247eece9701544a` | `conan create . binutils/2.38@ -o target_os=Linux`
| Linux    | Linux     | Windows     | `ac7cb2c348c6917e8ee2a45dd0268fedfc31d6d1` | `conan create . binutils/2.38@ -o target_os=Windows`
| Linux    | Windows   | Linux       | `8140b7dcc9c74a04fdcff4f118c8cc3b5dcbd5b4` | `conan create . binutils/2.38@ -pr:h mingw64 -pr:b default64 -o binutils:target_os=Linux`
| Linux    | Windows   | FreeBSD     | `5fbdaa86696ff852b8b301d8c15039a082ea75a5` | `conan create . binutils/2.38@ -pr:h mingw64 -pr:b default64 -o binutils:target_os=FreeBSD`

So all different host/target combinations result in different package ids, isn't that what we want?

If I'm wrong somewhere, can you please give me an explicit example?

**Comment by madebr** [comment]
>

**Comment by madebr** [comment]
> probably, it's not useful. let's disable it until someone actuallly requests.

Correction, this should be:

"let's disable it until someone ~actually requests it~ **fixes** it"

 :smiley:

**Comment by hlewin** [comment]
I would suggest giving the user a way to explicitely state the target triple/prefix for the package. Deriving it from settings_target will likely work for some cases but not for all.
Explanation: gcc and clang when cross-compiling will search for an assembler with an exactly identical target-prefix when invoked with `-fno-integrated-as` (which is needed to compile some programs). Unfortunately there is no way to know exactly what these prefixes will look like as there is some degree of freedom in the prefixes. For android-devices for example the official ndk uses prefixes like `aarch64-linux-android21`, encoding the device's API level in the prefix. (This is a case I personally would be interested in). But this is generally the case: The GNU convention https://www.gnu.org/software/autoconf/manual/autoconf-2.65/html_node/Specifying-Target-Triplets.html allows to put arbitrary vendors into the triple - it is perfectly fine to use something like `x86_64-MyCompary-linux` for a toolchain targeting special hosts.

**Comment by madebr** [comment]
> I would suggest giving the user a way to explicitely state the target triple/prefix for the package. Deriving it from settings_target will likely work for some cases but not for all. Explanation: gcc and clang when cross-compiling will search for an assembler with an exactly identical target-prefix when invoked with `-fno-integrated-as` (which is needed to compile some programs). Unfortunately there is no way to know exactly what these prefixes will look like as there is some degree of freedom in the prefixes. For android-devices for example the official ndk uses prefixes like `aarch64-linux-android21`, encoding the device's API level in the prefix. (This is a case I personally would be interested in). But this is generally the case: The GNU convention https://www.gnu.org/software/autoconf/manual/autoconf-2.65/html_node/Specifying-Target-Triplets.html allows to put arbitrary vendors into the triple - it is perfectly fine to use something like `x86_64-MyCompany-linux-gnu` for a toolchain targeting special hosts.

I've added an options to override the prefix.
I'm hesitant to add an option to override the gnu triplet as this would make the `target_{arch,os}` thing redundant and contradictory (as the triplet option would override `target_{arch,os}`)

**Comment by hlewin** [comment]
> I've added an options to override the prefix.
> I'm hesitant to add an option to override the gnu triplet as this would make the `target_{arch,os}` thing redundant and contradictory (as the triplet option would override `target_{arch,os}`)

I am unsure if specifying a different exec-prefix is sufficient - some arm targets have abis encoded in the last component (like arm-none-eabi). Dunno if this makes a difference for binutils but it seems kinda likely.

About the redundancy - yes, kinda. My approach would have been:
- store the triplet to be used as an option. This is used when building the toolchain.
- Set the triplet option from settings_target (when the package is used as build_requirement) and no triple-option is given
- Check for contradictions if settings_target and the triple-option is given

**Comment by hlewin** [comment]
@madebr I thank you for your efforts covering such cases. This way, I feel confident to being able to use this package for supplying binutils for arbitrary toolchains (which tend to add up recently).
Thanks

**Comment by madebr** [comment]
> @madebr I thank you for your efforts covering such cases. This way, I feel confident to being able to use this package for supplying binutils for arbitrary toolchains (which tend to add up recently). Thanks

Binutils is the easy part.
Combining it with glibc/gcc is the hard part. (It would be nice to have a recipe with a complete gcc toolchain)

**Comment by hlewin** [comment]
> Binutils is the easy part.
> Combining it with glibc/gcc is the hard part, certainly if you're doing a canadian cross build.
> (It would be nice to have a recipe for a complete gcc toolchain)

Well, at least this should work for supplying a GNU-as for the newer versions of the android-ndk (which do not include it anymore, which makes compiling some programs hard as the llvm-as is not fully compatible). I find tinkering with gcc or llvm quite bothering as builds tend to fail after 20+ minutes.... I may have a shot at a clang-toolchain at some time (which seems to be a little easier than gcc).
Thanks again

**Comment by madebr** [comment]
The recipe contains a lot of hardcoding to convert between os/arch and triplets.
I haven't tested android yet.

**Comment by hlewin** [comment]
> The recipe contains a lot of hardcoding to convert between os/arch and triplets.
> I haven't tested android yet.

Well, I guess one would have to use the `target_triplet` option even when using the package as build_requires as Android would require `os.api_level` to be set to generate the correct triple. Having a look at the os-specific sub-settings in `.conan/settings.yml` doing all conversion correct would be a HUGE endeavor.

**Comment by hlewin** [comment]
Reading the source I think I found an issue (did not try in practice yet, though).
For armv7 (only) android uses a triple of the form `armv7a-linux-androideabiXX` (XX=api level).
Note the `eabi` in the last part.

File: recipes/binutils/all/conanfile.py [hunk 3]
```diff
@@ -1,70 +1,512 @@
...
+    def configure(self):
+        if self.options.target_triplet == self._PLACEHOLDER_TEXT:
+            if self.options.target_arch == self._PLACEHOLDER_TEXT:
+                # If target triplet and target arch are not set, initialize it from the target settings
+                self.options.target_arch = str(self._settings_target.arch)
+            if self.options.target_os == self._PLACEHOLDER_TEXT:
+                # If target triplet and target os are not set, initialize it from the target settings
+                self.options.target_os = str(self._settings_target.os)
+            # Initialize the target_triplet from the target arch and target os
+            self.options.target_triplet = _GNUTriplet.from_archos(_ArchOs(arch=str(self.options.target_arch), os=str(self.options.target_os), extra=dict(self._settings_target.values_list))).triplet
```
Comment (madebr) [inline recipes/binutils/all/conanfile.py:75]: @hlewin 
`os.api_level` is passed here to `_ArchOs` (through the `extra` argument)
Actually, all properties of the target settings object are passed here.

File: recipes/binutils/all/conanfile.py [hunk 4]
```diff
@@ -1,70 +1,512 @@
...
+
+    def test_android_armv8(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="armv8", os="Android", extra={"os.api_level": "24"}), _GNUTriplet(machine="aarch64", vendor=None, os="linux", abi="android24"), "aarch64-linux-android24")
+
+    def test_android_x86(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="x86", os="Android", extra={"os.api_level": "16"}), _GNUTriplet(machine="i686", vendor=None, os="linux", abi="android16"), "i686-linux-android16")
+
+    def test_android_x86_64(self):
+        self._test_osarch_to_gnutriplet(_ArchOs(arch="x86_64", os="Android", extra={"os.api_level": "29"}), _GNUTriplet(machine="x86_64", vendor=None, os="linux", abi="android29"), "x86_64-linux-android29")
+        self.assertEqual(_ArchOs(arch="x86_64", os="Android", extra={"os.api_level": "25"}), _ArchOs.from_triplet(_GNUTriplet.from_text("x86_64-linux-android29")))
```
Comment (madebr) [inline recipes/binutils/all/conanfile.py:505]: Here, I test the correct conversion between an android conan os/arch and gnu triplet.

Comment (madebr) [reply recipes/binutils/all/conanfile.py:509]: I copied these triplets from the latest android ndk archive.

File: recipes/binutils/all/conanfile.py [hunk 5]
```diff
@@ -1,70 +1,513 @@
...
+            suggested_archos = _ArchOs.from_triplet(target_gnu_triplet)
+            raise ConanInvalidConfiguration(f"target_arch={target_archos.arch}/target_os={target_archos.os} is not compatible with {target_gnu_triplet.triplet}. Change target triplet to {suggested_gnu_triplet.triplet}, or change target_arch/target_os to {suggested_archos.arch}/{suggested_archos.os}.")
+
+        # Check, when used as build requirement in a cross build, whether the target arch/os agree
+        settings_target = getattr(self, "settings_target", None)
+        if settings_target is not None:
+            if self.options.target_arch != settings_target.arch:
+                raise ConanInvalidConfiguration(f"binutils:target_arch={self.options.target_arch} does not match target architecture={settings_target.arch}")
+            if self.options.target_os != settings_target.os:
+                raise ConanInvalidConfiguration(f"binutils:target_os={self.options.target_os} does not match target os={settings_target.os}")
```
Comment (jgsogo) [inline recipes/binutils/all/conanfile.py:120]: These are the key lines of this recipe 🎉

File: recipes/binutils/all/test_package/conanfile.py [hunk 6]
```diff
@@ -1,11 +1,113 @@
...
+
+    def _has_ld(self):
+        if self._target_os in ("Macos"):
+            return False
+        return True
 
     def test(self):
+        # Run selftest (conversion between conan os/arch <=> gnu triplet)
+        with tools.chdir(os.path.dirname(self.deps_user_info["binutils"].recipe_path)):
+            self.run(f"python -m unittest {os.path.basename(self.deps_user_info['binutils'].recipe_path)} --verbose")
```
Comment (jgsogo) [inline recipes/binutils/all/test_package/conanfile.py:88]: I agree testing these things is quite important. There is a lot of time spent there. I wonder if those test classes should be moved inside the `test_package/` instead of living in the root `conanfile.py`. Probably from those files we could import the root `conanfile` to reuse helper classes.

Of course, that could be a refactor for another PR.

**Review by jgsogo — COMMENTED** [review]
All I can say is 🙌 

I haven't had time to enter the details, but the recipe looks good and, if you have tested it locally for a few use cases and it works, it should be mostly good. Packaging a cross-compiler with Conan v1 is pretty hard and there might be lots of corner cases, but we need to try it and I think this is a very good effort. Congratulations!

File: recipes/binutils/all/test_package/conanfile.py (hunk 6 above)
Comment (madebr) [reply recipes/binutils/all/test_package/conanfile.py:88]: The self-test can be moved to the test package, if it becomes possible to export functions from a recipe to its dependencies.
In previous experiments, I could only export strings through `self.user_info`.

Comment (jgsogo) [reply recipes/binutils/all/test_package/conanfile.py:88]: I was not thinking about _exporting_ those functions, but about `from ..conanfile import XXXX` in `test_package/conanfile.py` (if it works, given we are using Py3).

Comment (madebr) [reply recipes/binutils/all/test_package/conanfile.py:88]: I did not choose that approach because I wanted to have the ability to test an exported recipe, not the recipe that happens to live in the cci repo.

So when you run the following command,
```
conan test recipes/binutils/all/test_package binutils/2.38@
```
you know the exported recipe is tested, not a combination of the exported recipe and the recipe in the cci file tree.

**Review by jgsogo — APPROVED** [review]
LGTM. We can improve step by step

**Review by SSE4 — APPROVED** [review]

File: recipes/binutils/all/test_package/conanfile.py (hunk 6 above)
Comment (madebr) [reply recipes/binutils/all/test_package/conanfile.py:88]: for completeness sake, you can still run the self-tests of the file in the cci file tree without having to export/build anything, by running:
```bash
python -m unittest recipes/binutils/all/conanfile.py
```
